KRFAnalysisPass/KRF.cpp

```diff
@@ -0,0 +1,296 @@
+#include "llvm/Pass.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/ADT/StringRef.h"
+#include <unordered_set>
+#include <string>
+
+using namespace llvm;
+
+cl::opt<bool> Json ("json", cl::desc("Print output in json format"));
+
+namespace {
+struct KRF : public ModulePass {
+  static char ID;
+  KRF() : ModulePass(ID) {}
+  const std::unordered_set<std::string> blacklisted_functions = {"accept",
+"access",
+"bind",
+"brk",
+"chdir",
+"chmod",
+"chown",
+"chroot",
+"clock_getres",
+"clock_gettime",
+"clock_nanosleep",
+"clock_settime",
+"clone",
+"close",
+"connect",
+"creat",
+"dup",
+"dup2",
+"faccessat",
+"fchdir",
+"fchmod",
+"fchmodat",
+"fchown",
+"fchownat",
+"fcntl",
+"fdatasync",
+"flock",
+"fork",
+"fstat",
+"fstatfs",
+"fsync",
+"ftruncate",
+"getcpu",
+"getcwd",
+"getpeername",
+"getpgid",
+"getpriority",
+"getresgid",
+"getresuid",
+"getsid",
+"getsockname",
+"getsockopt",
+"gettimeofday",
+"ioctl",
+"kill",
+"lchown",
+"link",
+"linkat",
+"listen",
+"lstat",
+"madvise",
+"mincore",
+"mkdir",
+"mknod",
+"mknodat",
+"mlock",
+"mlock2",
+"mlockall",
+"mmap_pgoff",
+"mount",
+"mprotect",
+"mq_getsetattr",
+"mq_notify",
+"mq_open",
+"mq_timedreceive",
+"mq_timedsend",
+"mq_unlink",
+"mremap",
+"msgctl",
+"msgget",
+"msgrcv",
+"msgsnd",
+"msync",
+"munlock",
+"munlockall",
+"munmap",
+"newfstatat",
+"open",
+"openat",
+"pipe",
+"read",
+"readlink",
+"readlinkat",
+"reboot",
+"recvfrom",
+"recvmsg",
+"rename",
+"renameat",
+"renameat2",
+"rmdir",
+"rt_sigaction",
+"rt_sigpending",
+"rt_sigprocmask",
+"rt_sigqueueinfo",
+"rt_sigsuspend",
+"rt_sigtimedwait",
+"sigaction",
+"sigpending",
+"sigprocmask",
+"sigsuspend",
+"sigtimedwait",
+"sched_get_priority_max",
+"sched_get_priority_min",
+"sched_getaffinity",
+"sched_getattr",
+"sched_getparam",
+"sched_getscheduler",
+"sched_rr_get_interval",
+"sched_setaffinity",
+"sched_setattr",
+"sched_setparam",
+"sched_setscheduler",
+"select",
+"semctl",
+"semget",
+"semop",
+"sendmsg",
+"sendto",
+"setdomainname",
+"setgid",
+"sethostname",
+"setpgid",
+"setpriority",
+"setregid",
+"setresgid",
+"setresuid",
+"setreuid",
+"setsid",
+"setsockopt",
+"settimeofday",
+"setuid",
+"shmat",
+"shmctl",
+"shmdt",
+"shmget",
+"shutdown",
+"sigaltstack",
+"socket",
+"socketpair",
+"stat",
+"statfs",
+"swapoff",
+"swapon",
+"symlink",
+"symlinkat",
+"syncfs",
+"sysfs",
+"syslog",
+"tgkill",
+"time",
+"timer_create",
+"timer_delete",
+"timer_getoverrun",
+"timer_gettime",
+"timer_settime",
+"timerfd_create",
+"timerfd_gettime",
+"timerfd_settime",
+"tkill",
+"truncate",
+"umount",
+"unlink",
+"unlinkat",
+"uselib",
+"ustat",
+"utime",
+"wait4",
+"waitid",
+"write",
+"syscall"
+  };
+					
+  bool runOnModule(Module &M) override {
+    json::OStream* J = (Json) ? new json::OStream(errs()) : NULL;
```

Comment:
Let's use a dedicated output instead of `stderr`. You can add a `cl::opt<std::string>` (with a default filename) to provide the output path.

Reply:
You might also have an easier time using the structural JSON API rather than the streaming one; you won't have to worry about correctly opening and closing scopes. If you switch to that, it would be fine to emit [JSONL](http://jsonlines.org/) instead of a single JSON object.